# Домашнє завдання до теми "Розширені методи обробки даних "

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import numpy as np
import pandas as pd
data_path = 'data_M3/supermarket_sales.csv'
df = pd.read_csv(data_path)

1. Відсортуйте дані за спаданням ціни товара ('Unit price') і виведіть 5 перших рядочків на екран.

In [2]:
df.sort_values('Unit price', ascending=False).head(5)

Invoice ID Branch       City Customer type Gender         Product line  \
983  148-41-7930      C  Naypyitaw        Normal   Male    Health and beauty   
122  219-22-9386      B   Mandalay        Member   Male    Sports and travel   
930  641-62-7288      B   Mandalay        Normal   Male   Home and lifestyle   
494  437-53-3084      B   Mandalay        Normal   Male  Fashion accessories   
283  667-92-0055      A     Yangon        Member   Male    Health and beauty   

     Unit price  Quantity  Tax 5%    Total       Date   Time      Payment  \
983       99.96         7  34.986  734.706  1/23/2019  10:33         Cash   
122       99.96         9  44.982  944.622   3/9/2019  17:26  Credit card   
930       99.92         6  29.976  629.496  3/24/2019  13:33      Ewallet   
494       99.89         2   9.989  209.769  2/26/2019  11:48      Ewallet   
283       99.83         6  29.949  628.929   3/4/2019  15:02      Ewallet   

       cogs  gross margin percentage  gross income  Rating  
983  699.72                 4.761905        34.986     6.1  
122  899.64                 4.761905        44.982     4.2  
930  599.52                 4.761905        29.976     7.1  
494  199.78                 4.761905         9.989     7.1  
283  598.98                 4.761905        29.949     8.5

2. Знайдіть інвойс (рядок в даних) в категорії товарів Electronic accessories, в якому був куплений товар з найбільшою кількістю і найбільшою ціною. Виведіть цей рядок даних на екран.

In [3]:
df[df['Product line']=='Electronic accessories'].sort_values(['Unit price', 'Quantity'], ascending=[False, False])[:1]

Invoice ID Branch      City Customer type  Gender  \
209  817-69-8206      B  Mandalay        Normal  Female   

               Product line  Unit price  Quantity   Tax 5%     Total  \
209  Electronic accessories       99.73         9  44.8785  942.4485   

         Date   Time      Payment    cogs  gross margin percentage  \
209  3/2/2019  19:42  Credit card  897.57                 4.761905   

     gross income  Rating  
209       44.8785     6.5

3.1. Створіть нову колонку в даних `Rating Category`, яка визначає категорію рейтингу: високий, середній чи низький за наступним правилом:
- якщо рейтинг вищий або рівний 8 - то він високий (`High`)
- якщо рейтинг від 6 до 8 не включаючи - то він середній (`Medium`)
- інакше - низький (`Low`).

In [4]:
def get_rating_category(val):
    if val >= 8:
        return "High"
    elif val >= 6 and val < 8:
        return "Medium"
    else:
        return "Low"

In [6]:
df['Rating Category'] = df.Rating.apply(get_rating_category)
df[['Rating', 'Rating Category']].head(10)

Rating Rating Category
0     9.1            High
1     9.6            High
2     7.4          Medium
3     8.4            High
4     5.3             Low
5     4.1             Low
6     5.8             Low
7     8.0            High
8     7.2          Medium
9     5.9             Low

3.2. Виведіть, який відсоток даних становить кожне зі значень категорій рейтингу.

In [7]:
df['Rating Category'].value_counts(normalize=True)

Rating Category
Medium    0.356
High      0.329
Low       0.315
Name: proportion, dtype: float64

4.1. Запустіть код нижче аби створити змінну `time`. Ця змінна містить час в форматі "година:хвилина". Напишіть функцію `get_hour`, яка буде зі змінної формату як `time` дістати лише години і конвертувати їх у тип `int`.

Запустіть написану функцію на змінній `time` та перевірте правильність роботи функції.

In [8]:
time = '11:28'

In [9]:
def get_hour(time):
    return int(time.split(':')[0])

In [10]:
get_hour(time)

11

4.2. З допомогою функції `get_hour` і метода `DataFrame.apply()` створіть нову колонку `Hour` в нашому наборі даних, яка буде містити години здійснення покупки, обчислені на основі часу здійснення покупки `Time`.

In [11]:
df['Hour'] = df.Time.apply(get_hour)

4.3. Виведіть кількість значень за кожною годиною покупки і дайте відповідь "О яких годинах найбільше здійснюють покупок покупці і чому на Ваш погляд саме о цих годинах? Чи це слогічно, що найбільше покупок саме в ці години?".

In [12]:
df.Hour.value_counts()

Hour
19    113
13    103
15    102
10    101
18     93
11     90
12     89
14     83
16     77
20     75
17     74
Name: count, dtype: int64

Найбільше покупок було о 19 (113), це час після робочого дня,
коли найчастіше основна маса людей, яка працює - здійснює покупки.
Високі показники о 13 та 15 годині, що схоже на час обідньої перерви - коли можливо дозволити собі зробити покупки.
Також і о 10 високий показник - можливо у даному регіоні магазини розпочинають робочий час саме о 10.

5. Виведіть середнє значення чеку (`Total`) для покупок в містах, наявних в даних, за допомогою функції `groupby`.

In [14]:
df.groupby('City')['Total'].mean()

City
Mandalay     319.872506
Naypyitaw    337.099715
Yangon       312.354031
Name: Total, dtype: float64

6. Виведіть кількість покупок, середнє, мінімальне та максимальне значення для загальної суми чеку (Total) в містах, наявних в даних, за допомогою функції groupby.
Зробіть висновки з цих показників та дайте рекомендації для мережі супермаркетів.

In [15]:
df.groupby('City')['Total'].agg(['count', 'mean', 'min', 'max']).round(2)

count    mean    min      max
City                                    
Mandalay     332  319.87  18.64  1022.49
Naypyitaw    328  337.10  10.68  1042.65
Yangon       340  312.35  12.69  1039.29

Висновки:
Naypyitaw - має найвищу сер. суму по чеку (337.10), це говорить про те, що в даному регіоні купуються дорожчі товари, більш платоспроможні покупці.
Рекомендації:
1) Для підтримання рівня продажів у місті Naypyitaw, потрібно заохочувати клієнтів, можливо впровадити спеціальну програму лояльності для "віп" покупців
2) У інших містах продажі та середній чек схожі по сумам, можливо потрібно провести заохочувальні кампанії, акції, у години найбільших покупок - запустити "хеппі ауерс".

7. Ми внесли деякі зміни в наш набір даних, тож, давайте запишемо результат у файл. Запишіть оновлену версію даних у файл `supermarket_sales_enriched.csv`. Зверніть увагу, аби не записати нічого зайвого, наприклад, індекси рядків нам в файлі не потрібні, оскільки вони стандартні.

In [17]:
df.to_csv('data_M3/supermarket_sales_enriched.csv', index=False)